src/System Application/Test/URI/src/UriBuilderQueryTest.Codeunit.al

```diff
@@ -314,18 +314,92 @@ codeunit 135072 "Uri Builder Query Test"
+            + '?%C3%A8=%C3%A9AddedNotEncoded&%25C3%25A8=%25C3%25A9AddedEncoded' // Added parameters (the one that was already encoded should be double encoded)
             + '&moreGarbledStuff=%26%2F%5C%27%22%3F%3F%3F%2520' // Ensure escape and special characters are encoded
             + '&%EA%B3%A0%EC%96%91%EC%9D%B4=%3A30%26%EA%B3%A0%EC%96%91%EC%9D%B4', // Ensure character that use 3 bytes are also correctly encoded
             Uri.GetAbsoluteUri(), 'Unexpected URL.');
     end;
 
+    [Test]
+    [Scope('OnPrem')]
+    procedure TestExpectedODataEncoding_ODataParameters()
+    var
+        Uri: Codeunit Uri;
+    begin
+        // [Given] A Url
+        UriBuilder.Init('https://microsoft.com?$top=33&%24skip=41&$filter=nothing&è=éNotEncoded&%C3%A8=%C3%A9Encoded');
+
+        // [When] Adding OData parameters that include the $ sign
+        UriBuilder.AddODataQueryParameter('$filter', 'Name eq ''&Contoso''');
+        UriBuilder.AddODataQueryParameter('$expand', 'Products($filter=DiscontinuedDate eq null)');
+        UriBuilder.AddODataQueryParameter('moreGarbledStuff😊', '&/\''"*!???%20');
+
+        // [Then] The resulting URI has encoded query parameters, except the $ sign in the parameter name
+        UriBuilder.GetUri(Uri);
+        Assert.AreEqual('https://microsoft.com/?$top=33&$skip=41' // Initial parameters
+            + '&$filter=Name%20eq%20%27%26Contoso%27' // Filter
+            + '&%C3%A8=%C3%A9NotEncoded&%C3%A8=%C3%A9Encoded' // Other initial parameters
+            + '&$expand=Products%28%24filter%3DDiscontinuedDate%20eq%20null%29' // $ is encoded in the value
+            + '&moreGarbledStuff%F0%9F%98%8A=%26%2F%5C%27%22%2A%21%3F%3F%3F%2520', // Non-OData parameter
+            Uri.GetAbsoluteUri(), 'Unexpected URL.');
+    end;
+
+    [Test]
+    [Scope('OnPrem')]
+    procedure TestExpectedODataEncoding_ODataParametersThenFlag()
+    var
+        Uri: Codeunit Uri;
+    begin
+        // [Given] A Url
+        UriBuilder.Init('https://microsoft.com?$top=33&%24skip=41&$filter=nothing&è=éNotEncoded&%C3%A8=%C3%A9Encoded');
+
+        // [When] Adding OData parameters that include the $ sign
+        UriBuilder.AddODataQueryParameter('$filter', 'Name eq ''&Contoso''');
+        UriBuilder.AddODataQueryParameter('$expand', 'Products($filter=DiscontinuedDate eq null)');
+        UriBuilder.AddODataQueryParameter('moreGarbledStuff😊', '&/\''"*!???%20');
+        UriBuilder.AddQueryFlag('$newschemaversion');
+
+        // [Then] The resulting URI has encoded query parameters, except the $ sign in the parameter name
+        UriBuilder.GetUri(Uri);
+        Assert.AreEqual('https://microsoft.com/?%24top=33&%24skip=41' // Initial parameters
+            + '&%24filter=Name%20eq%20%27%26Contoso%27' // Filter
+            + '&%C3%A8=%C3%A9NotEncoded&%C3%A8=%C3%A9Encoded' // Other initial parameters
+            + '&%24expand=Products%28%24filter%3DDiscontinuedDate%20eq%20null%29' // $ is encoded in the value
```

Comment:
Wouldn't you expect this to be &$expand? Just like in the previous test? After all you added "UriBuilder.AddODataQueryParameter('$expand', 'Products($filter=DiscontinuedDate eq null)');" the exact same way as above, hence I would expect the output to be the exact same.
I noticed the problem however is that the url is regenerated every time you add something new. If you add an OData Query parameter as the last, then ALL parameters will be formatted according to OData, if the last parameter is not like in this case, then all parameters are formatted as though they are not OData. This is because we convert the entire uri to a single string and then re-encode everything whenever you add a parameter.

Reply:
I also noticed the comments are the same as in the previous test, even though the result is different.